internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,368 @@
+		noLockWait = lockmanager.Lock(this.LockKey)
+		defer lockmanager.Unlock(this.LockKey)
+	}
+
+	// The context has been canceled while waiting for a lock, abandon this job.
+	if this.Context.Err() != nil {
+		return
+	}
+
+	var err error = nil
+
+	// If we had to wait for the lock, then check again for stored records.
+	if !noLockWait && this.RetrieveFunc != nil {
+		partialResults := []OutputType{}
+		remainingItems := []InputType{}
+		partialResults, remainingItems, err = this.RetrieveFunc(output.RemainingItems)
+		if err != nil {
+			output.Error = fmt.Errorf("Failed to re-check record storage before run: '%w'.", err)
+			return
+		}
+
+		output.RemainingItems = remainingItems
+
+		if updateOutput {
+			// Collect the partial records from storage.
+			output.ResultItems = append(output.ResultItems, partialResults...)
+		}
+	}
+
+	if len(output.RemainingItems) == 0 {
+		return
+	}
+
+	type PoolResult struct {
+		Input   InputType
+		Result  OutputType
+		RunTime int64
+		Error   error
+	}
+
+	poolResults, _, err := util.RunParallelPoolMap(this.PoolSize, output.RemainingItems, this.Context, func(workItem InputType) (PoolResult, error) {
+		workItemKey := ""
+		if this.WorkItemKeyFunc != nil {
+			workItemKey = this.WorkItemKeyFunc(workItem)
+		}
+
+		// Optionally lock this id so we don't work on an item multiple times.
+		if workItemKey != "" {
+			lockmanager.Lock(workItemKey)
+			defer lockmanager.Unlock(workItemKey)
+		}
+
+		if this.Context.Err() != nil {
+			return PoolResult{}, nil
+		}
+
+		// Check storage for a record.
+		if this.RetrieveFunc != nil {
+			results, _, err := this.RetrieveFunc([]InputType{workItem})
+			if err != nil {
+				return PoolResult{
+					Input: workItem,
+					Error: fmt.Errorf("Failed to retrieve item '%v': '%w'.", workItem, err),
+				}, nil
+			}
+
+			if len(results) == 1 {
+				return PoolResult{workItem, results[0], 0, nil}, nil
+			}
+		}
+
+		// Nothing cached, compute the job.
+		startTime := timestamp.Now()
+
+		result, err := this.WorkFunc(workItem)
+		if err != nil {
+			return PoolResult{
+				Input: workItem,
+				Error: fmt.Errorf("Failed to perform individual work on item '%v': '%w'.", workItem, err),
+			}, nil
+		}
+
+		runTime := (timestamp.Now() - startTime).ToMSecs()
+
+		// Store the result.
+		if !this.DryRun && this.Context.Err() == nil && this.StoreFunc != nil {
+			err = this.StoreFunc([]OutputType{result})
+			if err != nil {
+				return PoolResult{
+					Input: workItem,
+					Error: fmt.Errorf("Failed to store result for item '%v': '%w'.", workItem, err),
+				}, nil
+			}
+		}
+
+		return PoolResult{workItem, result, runTime, nil}, nil
+	})
+
+	if err != nil {
+		output.Error = fmt.Errorf("Failed to run job in a parallel pool: '%w'.", err)
+		return
+	}
+
+	// If the job was canceled, exit right away.
+	if this.Context.Err() != nil {
+		return
+	}
+
+	output.RemainingItems = []InputType{}
+
+	for _, poolResult := range poolResults {
+		if poolResult.Error != nil {
```

Comment:
These error semantics seem off to me.
 - Why are work errors merged with system errors? I could instead see like a single error that is like "there were work errors", then people can check for any errors, but not have system errors polluted.
 - Putting error items in remaining items seems a bit strange to me (I could see either leaving them out or having another field for them).